pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -28,43 +28,137 @@ import (
 }
 
 // NewWatcher provides a new watcher
-func NewWatcher(sockDir string) Watcher {
-	return Watcher{
-		path:     sockDir,
-		handlers: make(map[string]RegisterCallbackFn),
-		fs:       &utilfs.DefaultFs{},
+func NewWatcher(sockDir string) *Watcher {
+	return &Watcher{
+		path: sockDir,
+		fs:   &utilfs.DefaultFs{},
+
+		handlers:    make(map[string]PluginHandler),
+		plugins:     make(map[string]pathInfo),
+		pluginsPool: make(map[string]map[string]*sync.Mutex),
 	}
 }
 
-// AddHandler registers a callback to be invoked for a particular type of plugin
-func (w *Watcher) AddHandler(pluginType string, handlerCbkFn RegisterCallbackFn) {
+func (w *Watcher) AddHandler(pluginType string, handler PluginHandler) {
 	w.mutex.Lock()
 	defer w.mutex.Unlock()
-	w.handlers[pluginType] = handlerCbkFn
+
+	w.handlers[pluginType] = handler
 }
 
-// Creates the plugin directory, if it doesn't already exist.
-func (w *Watcher) createPluginDir() error {
+func (w *Watcher) getHandler(pluginType string) (PluginHandler, bool) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	h, ok := w.handlers[pluginType]
+	return h, ok
+}
+
+// Start watches for the creation of plugin sockets at the path
+func (w *Watcher) Start() error {
+	glog.V(2).Infof("Plugin Watcher Start at %s", w.path)
+	w.stopCh = make(chan interface{})
+
+	// Creating the directory to be watched if it doesn't exist yet,
+	// and walks through the directory to discover the existing plugins.
+	if err := w.init(); err != nil {
+		return err
+	}
+
+	fsWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin fsWatcher, err: %v", err)
+	}
+	w.fsWatcher = fsWatcher
+
+	w.wg.Add(1)
+	go func(fsWatcher *fsnotify.Watcher) {
+		defer w.wg.Done()
+		for {
+			select {
+			case event := <-fsWatcher.Events:
+				//TODO: Handle errors by taking corrective measures
+				go func() {
+					if event.Op&fsnotify.Create == fsnotify.Create {
+						err := w.handleCreateEvent(event)
+						if err != nil {
+							glog.Errorf("error %v when handling create event: %s", err, event)
+						}
+					} else if event.Op&fsnotify.Remove == fsnotify.Remove {
+						err := w.handleDeleteEvent(event)
+						if err != nil {
+							glog.Errorf("error %v when handling delete event: %s", err, event)
+						}
+					}
+					return
+				}()
+				continue
+			case err := <-fsWatcher.Errors:
+				if err != nil {
+					glog.Errorf("fsWatcher received error: %v", err)
+				}
+				continue
+			case <-w.stopCh:
+				fsWatcher.Close()
+				return
+			}
+		}
+	}(fsWatcher)
+
+	// Traverse plugin dir after starting the plugin processing goroutine
+	if err := w.traversePluginDir(w.path); err != nil {
+		fsWatcher.Close()
+		return fmt.Errorf("failed to traverse plugin socket path, err: %v", err)
+	}
+
+	return nil
+}
+
+// Stop stops probing the creation of plugin sockets at the path
+func (w *Watcher) Stop() error {
+	close(w.stopCh)
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		w.wg.Wait()
```

Comment:
Not sure that the use of the work group is making the synchronization any better... This code can be kept simpler by synchronizing around the `w.stopCh` only.

Reply:
This channel here is used to create a timeout on the call to Wait().

You seem to think that synchronizing around the `w.stopCh` would make the code simpler, how would I do that

Reply:
Repeated from above:
>In this context, it seems that once the processing loop (inside Start) is signaled to stop (via <-w.stopCh), there is no need for further coordination, there's nothing to wait for. Furthermore, any clean up can occur in fswatcher.Close().

I think in this case, once the processing loop is signaled to stop, it is safe to continue since there's nothing else to wait for.  If so, the workgroup can be removed.